### PR TITLE
FEAT: Basic datasets feature

### PR DESCRIPTION
**Pull request by codingl2k1**
## What do these changes do?

- [ ] ~~plugin merge chunks impl.~~  hard code for merging datasets in this PR.
  - [ ] ~~set initial_same_color_num = 1,~~ related issue: https://github.com/xorbitsai/xorbits/issues/384
- [x] spread loader op among nodes.
  - [ ] stable bind band to op for cache reuse. 
- [x] handle datasets without data files.
  - [ ] try download parquets if not data files.
- [x] impl `to_dataframe()`.
  - [x] fill dtypes and shape from builder info if possible.
- [ ] use iterative tiling to parallel map arrow files or arrow record batches.
- [x] auto execute adapter.
- [x] add ut.
- [ ] add doc.
- [x] refine code and comments.

Known issue:
- fetch data from worker node may be failed because the datasets could be a file mapping.
- rechunk is not fully supported.

## Related issue number

Fixes #xxxx

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass

### CODE REVIEW

File: python/xorbits/_mars/core/entity/output_types.py [hunk 1]
```diff
@@ -31,6 +31,7 @@ class OutputType(Enum):
     dataframe_groupby = 8
     series_groupby = 9
     df_or_series = 10
+    huggingface_dataset = 11
```
Comment (aresnow1) [inline python/xorbits/_mars/core/entity/output_types.py:34]: Or just rename to dataset?

File: python/xorbits/datasets/block.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .._mars.core.entity.objects import ObjectChunk, ObjectChunkData
+
+
+class BlockData(ObjectChunkData):
```
Comment (aresnow1) [inline python/xorbits/datasets/block.py:19]: To keep consistency with others, rename to DatasetChunkData?

File: python/xorbits/datasets/block.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    __slots__ = ()
+    type_name = "BlockData"
+
+
+class Block(ObjectChunk):
```
Comment (aresnow1) [inline python/xorbits/datasets/block.py:24]: DatasetChunk

File: python/xorbits/datasets/block.py (hunk 2 above)
Comment (codingl2k1) [reply python/xorbits/datasets/block.py]: Fixed. Thanks.

File: python/xorbits/datasets/block.py (hunk 3 above)
Comment (codingl2k1) [reply python/xorbits/datasets/block.py]: Fixed. Thanks.

File: python/xorbits/datasets/__init__.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .adapter import MARS_DATASET_CALLABLES, MARS_DATASET_TYPE, _install
+from .backends.huggingface.core import from_huggingface
+
+globals().update(MARS_DATASET_CALLABLES)
```
Comment (aresnow1) [inline python/xorbits/datasets/__init__.py:19]: Why need this line?

File: python/xorbits/datasets/backends/huggingface/loader.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            for index, (f, band) in enumerate(
+                zip(data_files, itertools.cycle(all_bands))
+            ):
+                chunk_op = op.copy().reset_key()
+                assert f, "Invalid data file from DatasetBuilder."
+                chunk_op.single_data_file = f
+                chunk_op.num_chunks = len(data_files)
+                chunk_op.chunk_index = index
+                chunk_op.expect_band = band
+                c = chunk_op.new_chunk(inputs=[], index=index)
```
Comment (aresnow1) [inline python/xorbits/datasets/backends/huggingface/loader.py:106]: Is index a tuple of int here? And chunk_op.chunk_index is not needed.

File: python/xorbits/datasets/__init__.py (hunk 4 above)
Comment (codingl2k1) [reply python/xorbits/datasets/__init__.py:19]: I need to update the wrapped callables to datasets module. e.g. a wrapped `from_huggingface`.

File: python/xorbits/datasets/backends/huggingface/loader.py (hunk 5 above)
Comment (codingl2k1) [reply python/xorbits/datasets/backends/huggingface/loader.py:106]: I am not sure if the index field can be empty. It's not a dataframe index.

Comment (aresnow1) [reply python/xorbits/datasets/backends/huggingface/loader.py:106]: For dataset chunk, indexes can be (0,), (1,) ...

Comment (codingl2k1) [reply python/xorbits/datasets/backends/huggingface/loader.py:106]: Got it. Thanks.

File: .github/workflows/python.yaml [hunk 6]
```diff
@@ -229,6 +230,10 @@ jobs:
           # test jax
           pip install "jax>=0.4.0" "jaxlib>=0.4.0"
         fi
+        if [[ "$MODULE" == "datasets" ]]; then
+          # test jax
```
[the remark this answers is not in the document]
Comment (codingl2k1) [reply .github/workflows/python.yaml:234]: Nice catch, thanks.

File: python/xorbits/_mars/utils.py [hunk 7]
```diff
@@ -152,6 +152,11 @@ def __repr__(self) -> str:
     except (ImportError, AttributeError):
         pass
 
+try:
+    import datasets as _hf_datasets
```
[the remark this answers is not in the document]
Comment (codingl2k1) [reply python/xorbits/_mars/utils.py]: Fixed. Thanks.

**Review by aresnow1 — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**